distribution/lib/Standard/Microsoft/0.0.0-dev/src/Internal/SQLServer_Type_Mapping.enso

```diff
@@ -0,0 +1,188 @@
+                    True  -> SQL_Type.Value Types.BINARY "VARBINARY(MAX)"
+                    False -> case variable_length of
+                        True  -> SQL_Type.Value Types.BINARY "VARBINARY" size
+                        False -> SQL_Type.Value Types.BINARY "BINARY" size
+            Value_Type.Mixed -> Error.throw (Unsupported_Database_Operation.Error "SQLServer tables do not support Mixed types.")
+            Value_Type.Unsupported_Data_Type type_name underlying_type ->
+                underlying_type.if_nothing <| Error.throw <| Illegal_Argument.Error <|
+                    "An unsupported SQL type ["+type_name.to_text+"] cannot be converted into an SQL type because it did not contain the SQL metadata needed to reconstruct it."
+
+        approximated_value_type = SQLServer_Type_Mapping.sql_type_to_value_type result
+        problems = if approximated_value_type == value_type then [] else [Inexact_Type_Coercion.Warning value_type approximated_value_type]
+        on_problems.attach_problems_before problems result
+
+    ## PRIVATE
+    sql_type_to_value_type : SQL_Type -> Value_Type
+    sql_type_to_value_type sql_type =
+        simple_type = case sql_type.typeid of
+            Types.BIT -> Value_Type.Boolean
+            Types.TINYINT -> Value_Type.Integer Bits.Bits_16
+            Types.SMALLINT -> Value_Type.Integer Bits.Bits_16
+            Types.INTEGER -> Value_Type.Integer Bits.Bits_32
+            Types.BIGINT -> Value_Type.Integer Bits.Bits_64
+            ## SQLServer real is 32-bit, but our In Memory engine only has 64-bit floating point numbers so we have to upcast it.
+            Types.REAL -> Value_Type.Float Bits.Bits_64
+            Types.DOUBLE -> Value_Type.Float Bits.Bits_64
+            Types.DATE -> Value_Type.Date
+            Types.TIME -> Value_Type.Time
+            Types.TIMESTAMP -> case sql_type.name of
+                "smalldatetime" -> Value_Type.Date_Time with_timezone=False
+                "datetime" -> Value_Type.Date_Time with_timezone=False
+                "datetime2" -> Value_Type.Date_Time with_timezone=False
+                _ -> on_unknown_type sql_type
+            Types.BINARY -> case sql_type.name of
+                "varbinary" -> Value_Type.Binary size=sql_type.precision variable_length=True
+                "binary" -> Value_Type.Binary size=sql_type.precision variable_length=False
+                _ -> on_unknown_type sql_type
+            _ -> case sql_type.name of
+                "datetimeoffset" -> Value_Type.Date_Time with_timezone=True
+                _ -> Nothing
+        simple_type.if_nothing <|
+            ## If we didn't match any of the types from the simple mapping, we
+               continue with the more complex mappings that take stuff like
+               precision into account.
+            case complex_types_map.get sql_type.typeid Nothing of
+                Nothing -> on_unknown_type sql_type
+                builder -> builder sql_type
```

Comment:
I think that with the `case-of` pattern match here, we no longer need to distinguish between the _"simple"_ and _"complex"_ mappings - we should probably merge all this into a big 'switch' statement. Before this was split only because we could not use the Java enum constants in the pattern match and needed various hacks to make the code readable.

Reply:
Done